x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,296 @@
+                TaskType.ANY,
+                inferenceId,
+                null,
+                List.of(query),
+                Map.of(),
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+            );
+
+            inferenceResultsSupplier = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    InferenceAction.INSTANCE,
+                    inferenceRequest,
+                    listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                        inferenceResultsSupplier.set(inferenceResponse.getResults());
+                        l.onResponse(null);
+                    })
+                )
+            );
+        } else {
+            // The most likely reason for a null inference ID is an invalid field name, invalid index name(s), or a combination of both.
+            // Set the inference results to an empty list so that query rewriting can continue.
+            // Invalid index names will be handled in TransportSearchAction, which will throw IndexNotFoundException.
+            // Invalid field names will be handled later in the rewrite process.
+            inferenceResultsSupplier = new SetOnce<>(new SparseEmbeddingResults(List.of()));
+        }
+
+        return new SemanticQueryBuilder(this, inferenceResultsSupplier, null);
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        throw new IllegalStateException(NAME + " should have been rewritten to another query type");
+    }
+
+    private static String getInferenceIdForForField(Collection<IndexMetadata> indexMetadataCollection, String fieldName) {
+        String inferenceId = null;
+        for (IndexMetadata indexMetadata : indexMetadataCollection) {
+            InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(fieldName);
+            String indexInferenceId = inferenceFieldMetadata != null ? inferenceFieldMetadata.getInferenceId() : null;
+            if (indexInferenceId != null) {
+                if (inferenceId != null && inferenceId.equals(indexInferenceId) == false) {
+                    throw new IllegalArgumentException("Field [" + fieldName + "] has multiple inference IDs associated with it");
+                }
+
+                inferenceId = indexInferenceId;
+            }
+        }
+
+        return inferenceId;
+    }
+
+    @Override
+    protected boolean doEquals(SemanticQueryBuilder other) {
+        return Objects.equals(fieldName, other.fieldName)
+            && Objects.equals(query, other.query)
+            && Objects.equals(inferenceResults, other.inferenceResults)
+            && inferenceResultsSuppliersEqual(inferenceResultsSupplier, other.inferenceResultsSupplier);
```

Comment:
```suggestion
            && Objects.equals(inferenceResults, other.inferenceResults);
```
Since its ephemeral, we shouldn't check it when doing equality.